compiler/x/codegen/X86FastCallLinkage.cpp

```diff
@@ -0,0 +1,240 @@
+
+#include "codegen/Machine.hpp"
+#include "codegen/Register.hpp"
+#include "codegen/RegisterDependency.hpp"
+#include "codegen/RegisterPair.hpp"
+#include "codegen/RealRegisterManager.hpp"
+#include "env/jittypes.h"
+#include "il/DataTypes.hpp"
+#include "il/Node.hpp"
+#include "il/Node_inlines.hpp"
+#include "il/TreeTop.hpp"
+#include "il/TreeTop_inlines.hpp"
+#include "x/codegen/X86Instruction.hpp"
+
+#ifdef TR_TARGET_64BIT
+#include "x/amd64/codegen/AMD64FastCallConfig.hpp"
+#else
+#include "x/i386/codegen/IA32FastCallConfig.hpp"
+#endif
+
+const size_t TR::X86FastCallCallSite::NumberOfIntParamRegisters   = sizeof(IntParamRegisters)   / sizeof(IntParamRegisters[0]);
+const size_t TR::X86FastCallCallSite::NumberOfFloatParamRegisters = sizeof(FloatParamRegisters) / sizeof(FloatParamRegisters[0]);
+const size_t TR::X86FastCallCallSite::StackIndexAdjustment        = RegisterParameterShadowOnStack ? NumberOfIntParamRegisters : 0;
+
+// Calculate preserved register map.
+// The map is a constant that the C++ compiler *should* be able to determine at compile time
+static uint32_t X86FastCallCallSiteCalculatePreservedRegisterMapForGC()
+   {
+   uint32_t ret = 0;
+   for (size_t i = 0;
+        i < sizeof(TR::X86FastCallCallSite::CalleeSavedRegisters) / sizeof(TR::X86FastCallCallSite::CalleeSavedRegisters[0]);
+        i++)
+      {
+      ret |= TR::RealRegister::gprMask(TR::X86FastCallCallSite::CalleeSavedRegisters[i]);
+      }
+   return ret;
+   }
+const uint32_t TR::X86FastCallCallSite::PreservedRegisterMapForGC = X86FastCallCallSiteCalculatePreservedRegisterMapForGC();
+
+TR::Register* TR::X86FastCallCallSite::BuildCall()
+   {
+   TR_ASSERT(CalleeCleanup || cg()->getLinkage()->getProperties().getReservesOutgoingArgsInPrologue(),
+             "Caller must reserve outgoing args in prologue unless callee cleans up stack");
+
+   if (cg()->comp()->getOption(TR_TraceCG))
+      {
+      traceMsg(cg()->comp(), "X86 FastCall: [%04d] %s\n", _SymRef->getReferenceNumber(), cg()->getDebug()->getName(_SymRef));
+      }
+   RealRegisterManager RealRegisters(cg());
+   TR::RealRegister*   ESP = cg()->machine()->getX86RealRegister(TR::RealRegister::esp);
+
+   // Preserve caller saved registers
+   for (size_t i = 0;
+        i < sizeof(CallerSavedRegisters) / sizeof(CallerSavedRegisters[0]);
+        i++)
+      {
+      RealRegisters.use(CallerSavedRegisters[i]);
+      }
+   // Find the return register, EAX/RAX/XMM0
+   TR::Register* EAX = RealRegisters.use(TR::RealRegister::eax);
+   TR::Register* XMM0 = RealRegisters.use(TR::RealRegister::xmm0);
+
+   // Build parameters, parameters in _Params are Right-to-Left (RTL)
+   TR::Register* IntParams[NumberOfIntParamRegisters] = {};
+   TR::Register* FloatParams[NumberOfFloatParamRegisters] = {};
+   TR_Stack<TR::Register*> StackParams(cg()->trMemory());
+
+   size_t IntParamIndex = 0;
+#ifndef WINDOWS
+   size_t FloatParamIndex = 0;
+#else
+   auto& FloatParamIndex = IntParamIndex;
```

Comment:
I'm not a big fan of this construct. While it moves the #ifdef out of the loop, it makes that loop code somewhat harder to understand (the parameter indices could be the same variable or not, depending on the platform). This is an example where I think I'd actually prefer to see the #ifdef code inside the loop (gasp! I know...)

Reply:
I am still not a huge fan, but I can't argue that it doesn't match how the Windows ABI maps registers. I'll withdraw the opinion as not important enough to block the change.